File: ios/Toodle/Rust/List/Items/Item.swift
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import Foundation

class Item {
    var raw: OpaquePointer?
    var id: Int64?
    var uuid: UUID?
    var name: String
    var dueDate: Date? {
        get {
            return ToodleLib.sharedInstance.value(forAttribute: ":todo/due_date", onEntity: self.id!)?.asDate()
        }
        set {
            if let date = newValue {
                store_set_timestamp_for_attribute_on_entid(ToodleLib.sharedInstance.intoRaw(), self.id!, ":todo/due_date", date.asInt64Timestamp())
            }
        }
    }
    var completionDate: Date? {
        get {
            return ToodleLib.sharedInstance.value(forAttribute: ":todo/completion_date", onEntity: self.id!)?.asDate()
        }
        set {
            if let date = newValue {
                store_set_timestamp_for_attribute_on_entid(ToodleLib.sharedInstance.intoRaw(), self.id!, ":todo/completion_date", date.asInt64Timestamp())
            }
        }
    }

    fileprivate var _labels: [Label]?

    init(name: String) {
        self.name = name
    }

    init(id: Int64, uuid: UUID, name: String) {
        self.name = name
        self.uuid = uuid
        self.id = id
    }

    var labels: [Label] {
        get {
            if _labels == nil {
                _labels = []
                // TODO: When we get labels in, put this back!
//                let ls = item_get_labels(self.raw)
//                _labels = []
//                for index in 0..<item_labels_count(ls) {
//                    let label = Label(raw: item_label_at(ls, index)!)
//                    _labels?.append(label)
//                }
            }

            return _labels!
        }
        set {
            _labels = nil
        }
    }

    func dueDateAsString() -> String? {
        guard let dueDate = self.dueDate else {
            return nil
        }
        let dateFormatter = DateFormatter()
        dateFormatter.dateStyle = .long
        dateFormatter.timeStyle = .short
        return dateFormatter.string(from: dueDate)
    }

    func completionDateAsString() -> String? {
        guard let completionDate = self.completionDate else {
            return nil
        }
        let dateFormatter = DateFormatter()
        dateFormatter.dateStyle = .long
        dateFormatter.timeStyle = .short
        return dateFormatter.string(from: completionDate)
    }
}

File: android/Toodle/app/src/main/java/com/mozilla/toodle/Item.java
/* -*- Mode: Java; c-basic-offset: 4; tab-width: 20; indent-tabs-mode: nil; -*-
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package com.mozilla.toodle;

import android.content.Context;

import com.mozilla.toodle.rust.NativeItem;
import com.mozilla.toodle.rust.Toodle;
import com.mozilla.toodle.rust.TypedValue;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.UUID;
import java.util.Date;

public class Item {
    private long id;
    private UUID uuid;
    private String name;
    private Date dueDate;
    private Date completionDate;

    public String name() {
        return name;
    }

    Item name(final String name) {
        this.name = name;
        return this;
    }

    Item() {

    }

    public Item(long id, UUID uuid, String name) {
        this.id = id;
        this.uuid = uuid;
        this.name = name;
    }

    public UUID uuid() {
        return uuid;
    }

    public Date dueDate() {
        if (this.dueDate == null) {
            TypedValue value = Toodle.getSharedInstance().valueForAttributeOnEntity(":todo/due_date", this.id);
            if (value != null) {
                this.dueDate = value.asDate();
            }
        }
        return this.dueDate;
    }

    public Date completionDate() {
        if (this.completionDate == null) {
            TypedValue value = Toodle.getSharedInstance().valueForAttributeOnEntity(":todo/completion_date", this.id);
            if (value != null) {
                this.completionDate = value.asDate();
            }
        }
        return this.completionDate;
    }

    Item completionDate(Long timestamp) {
        if (timestamp == null) {
            completionDate = null;
        } else {
            completionDate = new Date(timestamp / 1000);
        }
        return this;
    }

    Item dueDate(final int year, final int month, final int date) {
        final Calendar cal = Calendar.getInstance();
        cal.set(year, month, date);
        dueDate = new Date(cal.getTimeInMillis());
        return this;
    }

    private static Item fromNativeItem(NativeItem nativeItem) {
        final Item item = new Item();
        item.uuid = UUID.fromString(nativeItem.uuid);
        item.name = nativeItem.itemName;
        if (nativeItem.dueDate != null) {
            long timestamp = nativeItem.dueDate.getValue().longValue();
            if (timestamp == 0) {
                item.dueDate = null;
            } else {
                item.dueDate = new Date(timestamp);
            }
        }
        if (nativeItem.completionDate != null) {
            long timestamp = nativeItem.completionDate.getValue().longValue();
            if (timestamp == 0) {
                item.completionDate = null;
            } else {
                item.completionDate = new Date(timestamp);
            }
        }
        return item;
    }

    static ArrayList<Item> fromNativeItems(List<NativeItem> nativeItems) {
        final ArrayList<Item> items = new ArrayList<>(nativeItems.size());

        for (NativeItem nativeItem : nativeItems) {
            items.add(fromNativeItem(nativeItem));
        }

        return items;
    }

    void create(Context context) {
        Toodle.getSharedInstance(context).createItem(this);
    }

    void update(Context context) {
        Toodle.getSharedInstance(context).updateItem(this);
    }
}